### ISSUE TITLE
ZIM backend : Try to cross-compile the libZIM library in webassembly with emscripten

### ISSUE THREAD

**Issue by mossroy** [kiwix/kiwix-js#116]
It would avoid to re-code it in javascript, and would also ease support of future evolutions of the file format.
But not sure how Emscripten would handle the file I/Os

**Comment by peter-x** [comment]
Being able to use libZIM as is depends on the question whether we can use synchronous IO at the low-level javascript side.
This probably has to wait until we know how we can access the filesystem from ServiceWorkers.

**Comment by mossroy** [comment]
I don't think there is a way to do synchronous I/O on files in javascript.
Regarding ServiceWorkers, I suppose we have access to the same APIs as in any javascript.

**Comment by mossroy** [comment]
Dattaz, I assign you this issue, as you expressed some interest in it.
No hurry and no obligation of course

**Comment by mossroy** [comment]
@thiolliere if you want to have a look, too

**Comment by mossroy** [comment]
I met @bnjbvr yesterday (he was giving a conference about WebAssembly, at @mixitconf), and talked to him about this idea. He said he might put us in touch with the maintainer of emscripten if necessary.
Obviously, it's too early for now.

I found this article that tackles our issue : https://hacks.mozilla.org/2015/02/synchronous-execution-and-filesystem-access-in-emscripten/
There seems to be a notion of "virtual filesystem" that allows synchronous access to files that are preloaded in memory. It's clearly not possible for us because of the size of our ZIM files.
Refactoring the zimlib code to make asynchronous file I/O is probably complicated. It's a change that would only be interesting for us (not for the other applications using this library) and would probably make the code less readable.
Maybe the Emterpreter might be an option, but they say it would be slower. So maybe we would have to find a way to use the Emterpreter only on the code that reads the ZIM file, and compile everything else in WebAssembly, if it's technically possible.

**Comment by dattaz** [comment]
I have compile libzim in webassembly ; you can check demo here : https://dattaz.github.io/libzim_wasm/ (but zimfile (ekopedia_fr_all_2017-04.zim 12M) is embedded ) ; Now we have to deal with filesystem :) 

Source code to build is here : https://github.com/dattaz/libzim_wasm

**Comment by kelson42** [comment]
@dattaz  I get only - with FF 53 - a a page with black squares. Nothing interesting is displayed.

**Comment by mossroy** [comment]
Yeah! @dattaz, you rock!
It works pretty well, that's promising.
As you said, the next challenge is to deal with file I/Os. It might be the most difficult part, and might even not be technically possible.

The virtual filesystem you used for the demo can not work with bigger ZIM files that would not fit into memory.
We're left with the options in previous comment : 
- refactor libzim to make asynchronous file I/Os : it's the recommended way, and the best option for performance, but I don't know if it's not too complicated to do that
- or use Emterpreter to keep synchronous file I/Os : it would be much slower, according to Mozilla
In both cases, we have to make it use a javascript File object that we would pass to the wasm code : I hope it's possible

**Comment by dattaz** [comment]
Emscripten has WORKERFS as file system which permit to load file object as file into FS. This is only allowed in web worker. According doc (https://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html) it's seem really close of that we want to do : "This file system provides read-only access to File and Blob objects inside a worker without copying the entire data into memory and can potentially be used for huge files."

Here is a (little) demo : https://dattaz.github.io/libzim_wasm/file_api/index.html

Note that for the moment there is a issue with file bigger than 2GB : https://github.com/kripken/emscripten/issues/5250

**Comment by mossroy** [comment]
I did not know about workerfs : it looks great for our need!
This is very promising : if we manage to expose all the libzim functions through javascript, it might replace the low-level zim javascript code. It would be a huge improvement for kiwix-html5

**Comment by Jaifroid** [comment]
This page: https://github.com/kripken/emscripten/wiki/WebAssembly says that it's easy for projects to target both WebAssembly and asm.js. It would be good to keep the ASM option open not least because of the >2Gb file bug mentioned above, but also so that older browsers and packaged apps might benefit from any improvements by using libzim or kiwixzim over the current xzdec, which has some serious problems. Of course we don't know if libzim won't have similar problems.

I don't have a grasp on how File Systems work in Enscripten, (discussed [here]( https://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html#filesystem-api-filesystems)) but if xzdec.js is able to access enormous files, there must be hope for libzim.

**Comment by mossroy** [comment]
I fully agree with you : if Emscripten allows to easily compile in both WebAssembly and asm.js (and it seems to), we should try to do both. And provide asm.js as a fallback for the browsers that do not support WebAssembly.
But it needs to be tested...

Regarding large files, there's a big difference between xzdec and libzim : xzdec does not read any file (it takes a compressed stream as an input, and gives a decompressed one as the output), where libzim has to read the ZIM file(s) by itself.

**Comment by mossroy** [comment]
The libzim has evolved a lot since last year. It now relies on ninja + meson to build.
It's pretty easy to build it with the instructions from https://github.com/openzim/libzim.
But now we need to adapt it to compile with emscripten.

I managed to make it spit a build/src/libzim.so.4.0.4 that `file` recognizes as "LLVM IR bitcode".
That looks like a good first step to me.
I'm not very proud of how I did it, it's probably not the best way : 
I modified the generated build/build.ninja to replace cc by emcc and c++ by em++, and also had to modify src/debug.h to remove the dependency on execinfo.h and backtrace. I also manually copied the dependencies (lzma, xapian, zlib, zconf, and unicode) in build/include from /usr/include (it's probably not the right way to do it, but it seemed to work). Finally, I had to remove the warning about ABI version in xapian/version.h.
It does not manage to generate libzim.so.4.0.4.symbols, I hope it's not needed in our case?

Next step is to try to build the source code of the prototype https://github.com/dattaz/libzim_wasm with this new version of libzim, and check that it works the same. Before trying to plug it with kiwix-js

**Comment by mossroy** [comment]
Things are going forward, but it's still a long journey. I now need to compile the dependencies.
For the record : 
The ninja build gives warnings about these dependencies if I don't compile them :

> WARNING:root:emcc: cannot find library "lzma"
> WARNING:root:emcc: cannot find library "z"
> WARNING:root:emcc: cannot find library "xapian"
> WARNING:root:emcc: cannot find library "icui18n"
> WARNING:root:emcc: cannot find library "icuuc"
> WARNING:root:emcc: cannot find library "icudata"

I also managed to compile demo_file_api.cpp to wasm with : 
`em++ ../../libzim_wasm/demo_file_api.cpp -I../../libzim_wasm -Iinclude -I../include -fdiagnostics-color=always -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -O0 -g -D_LARGEFILE64_SOURCE=1 -D_FILE_OFFSET_BITS=64 -pthread --pre-js ../../libzim_wasm/prejs_file_api.js --post-js ../../libzim_wasm/postjs_file_api.js`
But with these warnings : 

> warning: unresolved symbol: _ZN3zim4FileC1ERKNSt3__212basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE
> warning: unresolved symbol: _ZNK3zim4File10getArticleEj
> warning: unresolved symbol: _ZNK3zim4File11getFilesizeEv
> warning: unresolved symbol: _ZNK3zim4File16getCountArticlesEv
> warning: unresolved symbol: _ZNK3zim4File17getArticleByTitleEj
> warning: unresolved symbol: _ZNK3zim4File3endEv
> warning: unresolved symbol: _ZNK3zim4File5beginEv
> warning: unresolved symbol: _ZNK3zim7Article6getUrlEv
> warning: unresolved symbol: _ZNK3zim7Article8getTitleEv

Then the prototype manages to start, runs some wasm code, but then fails with this error message : 

> missing function: _ZN3zim4FileC1ERKNSt3__212basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE

I'll try to compile the dependencies with emscripten one by one, to hopefully fix these problems

**Comment by mossroy** [comment]
On ICU, I had the following error : 

> Target architecture was not detected as supported by Double-Conversion.

Quick workaround : change line 101 of i18n/double-conversion-utils.h to replace the error by : 
`#define DOUBLE_CONVERSION_CORRECT_DOUBLE_OPERATIONS 1`
(hoping the emscripten architecture does not have the rounding issue on double that is mentioned in the comments of this source file)

At the end, it fails on running bin/icupkg, but that might not be necessary in our context.

**Comment by Jaifroid** [comment]
Bon courage! It'll be a game changer if you can get this working.

**Comment by mossroy** [comment]
To compile xapian : 
`emconfigure ./configure --prefix=`pwd`/../xapian "CFLAGS=-I`pwd`/../z/include -L`pwd`/../z/lib" "CXXFLAGS=-I`pwd`/../z/include -L`pwd`/../z/lib" --disable-backend-remote
emmake make "CFLAGS=-I`pwd`/../z/include -L`pwd`/../z/lib -std=c++11" "CXXFLAGS=-I`pwd`/../z/include -L`pwd`/../z/lib -std=c++11"`

I managed to compile the prototype without warnings with : 
`em++ ../../libzim_wasm/demo_file_api.cpp src/libzim.so -I../../libzim_wasm -Iinclude -I../include -fdiagnostics-color=always -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -O0 -g -D_LARGEFILE64_SOURCE=1 -D_FILE_OFFSET_BITS=64 -pthread --pre-js ../../libzim_wasm/prejs_file_api.js --post-js ../../libzim_wasm/postjs_file_api.js -s DISABLE_EXCEPTION_CATCHING=0 -s "EXTRA_EXPORTED_RUNTIME_METHODS=['ALLOC_NORMAL','printErr','ALLOC_STACK','ALLOC_STATIC','ALLOC_DYNAMIC','ALLOC_NONE']" -s DEMANGLE_SUPPORT=1`

It starts, but then crashes when it starts reading the ZIM file : 
> will print first 100 url/title of /work/wikipedia_en_ray_charles_2015-06.zim
> error reading zim-file header.

**Comment by mossroy** [comment]
I disabled MMAP in a very dirty way, by removing what's inside the ifdef of src/file_reader.cpp (it's certainly possible to do that in a cleaner way with the ENABLE_USE_MMAP option).

... and the prototype works!

**Comment by mossroy** [comment]
I needed to add the same `-s TOTAL_MEMORY=83886080` compilation option for the last compilation step (with em++), as we did for the xzdec compilation.

Good news : if I also add `-s WASM=0` compilation option, it is generated in asm.js instead of wasm, and also works. So it might be compatible with older browsers that do not support WASM

Bad news : big ZIM files still fail (apparently >2GB), with various error messages : 

> last cluster offset larger than file size; file corrupt

or

> Checksum position is not valid

Same behavior on Firefox and Chromium. So I suppose https://github.com/kripken/emscripten/issues/5250 has not been fixed since last year

**Comment by Jaifroid** [comment]
Still very encouraging results!

Do you know what File System emulation it's using? The problem @dattaz reported was with WORKERFS, but our current xzdec can be explicitly compiled with no filesystem: `NO_FILESYSTEM=1`.

I realize libzim is very different, but might it be made to accept a FileReader array buffer? Or must it work with an emulated file system?

**Comment by mossroy** [comment]
It is still WORKERFS, with good reasons : while xzdec takes a byte array as its input (because it only handles decompression), libzim actually reads the files (because it handles much more things).
Among the "filesystems" supported by emscripten (https://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html), I don't see another one that could fit our needs (using a javascript File object)

**Comment by Jaifroid** [comment]
I'm clutching at small straws here, but what about:
[Synchronous Virtual XHR Backed File System](https://kripken.github.io/emscripten-site/docs/porting/files/Synchronous-Virtual-XHR-Backed-File-System-Usage.html) ?

In particular this:

> The backend can improve start up time as the whole file system does not need to be preloaded before compiled code is run. It can also be very efficient if the web server supports byte serving — in this case Emscripten can just read the parts of files that are actually needed.

Perhaps this is the technology behind WORKERFS.

I guess the 2GB limitation is probably a version of https://en.wikipedia.org/wiki/2_GB_limit ? We might be back to using split ZIMs...

**Comment by mossroy** [comment]
Next step is probably to check we can call the libzim APIs from our javascript code. See http://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#interacting-with-code-binding-cpp
Maybe it will involve adding C++ code that wraps what we need into simple APIs that could be more easily called by javascript.

After discussing with other people at the hackathon, I understand that libzim is the low-level reference implementation to read ZIM files. It might be enough to replace most of our current javascript backend pieces, but https://github.com/kiwix/kiwix-lib provides more high-level APIs that should be more efficient for us. But it would need to compile it too, with its own dependencies. Hopefully we can do that later.

Our need would be to replace the calls to functions of zimArchive.js (the ones called by app.js) by some equivalent APIs running in wasm/asm.js
It might involve refactoring our javascript code in app.js.

One idea to test that might be to replace the calls made by handleMessageChannelMessage in app.js (getDirEntryByTitle, resolveRedirect and readBinaryFile) by a single call to getArticleByUrl (from libzim). It would not handle redirects properly (especially in stackoverflow ZIM files), but it would be an interesting step to see if it can work.

**Comment by mossroy** [comment]
The recompiled prototype (in its current form) can be tested on https://mossroy.github.io/libzim_wasm/ (it is the wasm version)

**Comment by Jaifroid** [comment]
> One idea to test that might be to replace the calls made by handleMessageChannelMessage in app.js (getDirEntryByTitle, resolveRedirect and readBinaryFile) by a single call to getArticleByUrl (from libzim). It would not handle redirects properly (especially in stackoverflow ZIM files), but it would be an interesting step to see if it can work.

This sounds do-able, as an experiment. The important thing is for us to build a clearly understandable pipeline / process for requesting articles. Anything that simplifies the micro-management of checking dirEntries and decompressing would be great (so long as it is performant).

> The recompiled prototype (in its current form) can be tested on https://mossroy.github.io/libzim_wasm/ (it is the wasm version).

I've tested on Edge with the same results that you report for the other two browsers. It does not, of course, work in IE, or in Firefox ESR 52.5.2 (the only version I could get that still supports the FFOS simulator). The latter spits out "No WebAssembly support found". But building an asm version would solve that (and as we have seen, may be just as performant).

The 2GB limit on files seen by the virtual file system limit (if that's what it is) might be a blocker unless it allows for split ZIMs. On the other hand, Kiwix seems to be phasing out pre-built split ZIMs. I guess the code would still be useful for packaged ZIMs < 2GB.

**Comment by mossroy** [comment]
I just pushed a new version that measures the execution time of reading an article from its URL.
It seems way faster than our current implementation. At least twice or 3 times faster, sometimes much more (most articles are read from the Ray Charles archive in 1 to 3 milliseconds, and from a 1.2GB ZIM file in around 70 milliseconds). It might come from the fact that the libzim has some built-in cache by itself.
It has to be verified in real conditions (because the content still needs to be transferred in javascript), but that might allow us to avoid adding other kind of caches like #411, #414 or #415. My personal opinion would be to put on hold the work on these custom cache implementations for now, until we find out if they will still be necessary.

**Comment by Jaifroid** [comment]
Excellent! I'm not working more on my cache code -- I did that in May, and only updated it to latest master since it was an outstanding issue.

I guess we still need to investigate the 2GB  file limit issue, as this was a blocker last year.

**Comment by mossroy** [comment]
Another good news is that I have similar performance with asm.js (see branch https://github.com/mossroy/libzim_wasm/tree/compiled-in-asm.js) IF I add the `-O3` parameter at compilation time (else it is very slow).
I tried to put this same `-O3` parameter in wasm version : it does not seem to have a significant impact on performance, but it makes generated files around one third smaller. I will keep `-O0` while we're still testing.

**Comment by mossroy** [comment]
I'm trying to call the C code from javascript, but it's not easy.
I managed (easily) to create some bindings using http://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/WebIDL-Binder.html#webidl-binder for File.getArticleByUrl that look good, but I don't manage to call it.
The generated a.out.js fails on load with : 

> ReferenceError: addOnPreMain is not defined

If I remove this call to addOnPreMain, it loads but fails when I call the constructor of File object with : 

> ReferenceError: assert is not defined

or 

> ReferenceError: intArrayFromString is not defined

**Comment by mossroy** [comment]
It seems to work a bit better with embind, and is not more complicated to setup if we wrap everything into high-level functions that don't use classes. I created a new branch for it.
But there might be issues with size of returned content, that we will have to work on, as it says : 

> Assertion failed: provided buffer should be 16777216 bytes, but it is 134217728

**Comment by mossroy** [comment]
After giving a `-s TOTAL_MEMORY=83886080` to the compiler (instead of letting the memory grow), this error disappears.
I've managed, in branch https://github.com/mossroy/libzim_wasm/tree/embind-experiments, to read an HTML content with libzim, and get it in javascript. That's cool!
I just pushed it to gh-pages so that it can be tested on https://mossroy.github.io/libzim_wasm/ (only on Ray Charles archive, and you have to wait for the "libzim initialized" message in the browser log before clicking on the second button)

**Comment by mossroy** [comment]
I've created a branch on kiwix-js to experiment with using libzim : https://github.com/kiwix/kiwix-js/tree/libzim-experiments
It's only a rough prototype to see how far we can get.
I only plugged it for the first HTTP request (because it only works once, for now), and only in SW mode. There is an unhandled race condition : the libzim needs to be initialized before it can be used : that's why I removed the automatic opening of the main article (you have to wait a bit before opening it).
Like for the libzim_wasm prototype, it only works on ZIM files of size <2GB.
And there is another issue : on some ZIM files, the HTML string returned by libzim seems to be a concatenation of many articles instead of only the expected article (making the browser struggle to handle everything), as if it was continuing to read the ZIM file content after the article end.
I also only implemented a simple C call (getArticleByUrl) that only handles strings : if we have to handle classes, it might be more difficult.

So there are still a lot of issues to handle, but it gives hope that there might a way to make this work in the future.

**Comment by mossroy** [comment]
Using split ZIM would probably be a workaround, but I could not check for now.
Regarding this 2GB limit, we got an answer on https://github.com/kripken/emscripten/issues/5250 , I'll try to ask for more details

**Comment by mossroy** [comment]
I've commited a Dockerfile to enable compilation of the libzim inside a Docker container (with emscripten and all the necessary tools).
I've also upgraded the versions in Makefile, to compile version 4.0.5 of libzim instead of 4.0.4
The HTML test page has also been slightly improved.

**Comment by kelson42** [comment]
@mossroy That sounds like an important step forward, but I'm not sure to understand concretly what it means. Doe it means the work to bind Kiwix JS with libzim-js could start?

BTW, I use the opportunity to inform you formally that @ISNIT0 has updated `node-libzim` https://github.com/openzim/node-libzim the nodejs binding to the libzim... and that mwoffliner 1.8 - to be release soon, use it.

**Comment by ISNIT0** [comment]
The bindings are written in nbind, which natively supports compiling to webassembly, possible to do :)

**Comment by mossroy** [comment]
No, what I did yesterday is not a significant step. It just simplifies reproducing what was achieved at the last hackathon on this topic (based on the work of dattaz).
We currently manage to compile libzim with emscripten, but using it in kiwix-js still needs a lot of work.
The current prototype is "only" a promising proof-of-concept.

There is at least one blocking issue : https://github.com/emscripten-core/emscripten/issues/5250 , which prevents reading files bigger than 2GB. This has to be fixed on the emscripten side, and I'm not able to do that myself. This issue has been tagged by kripken as "help wanted" : if you know someone able to work on that, it would be cool.

And there are also many things that need to be done before really using it in kiwix-js (see https://github.com/kiwix/kiwix-js/issues/116#issuecomment-423801066 above) : 
- compile kiwix-lib with emscripten, and use it instead of libzim, in order to benefit from its higher-level APIs
- replace the quick-and-dirty patches that I had to put on icu and libzim (see files patch_* in https://github.com/mossroy/libzim_wasm) by more sustainable fixes
- test how emscripten lets us use C APIs that have more complicated signatures than strings : how can we pass and/or use return values that are arrays or C classes? I'm currently using embind for these bindings, which should let us do that : https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html
- compare the APIs of our current backend with the ones of kiwix-lib, and see how we can use them. It will certainly not be a perfect match, and we probably need to add C and/or javascript glue
- fix the fact that we currently can do only one call to the C API, and properly wait for the wasm binary to be initialized before using it
- check that it works with split ZIM files
- check if asm.js can still be used as a fallback for platforms where webassembly is not available
- and probably other issues that we will discover

All this is very exciting. I'm just missing some time to work on it.

**Comment by mossroy** [comment]
Regarding https://github.com/openzim/node-libzim, I currently don't see how it could be useful for kiwix-js.
Maybe I misunderstood but I don't think it can be used in a browser environment, as it downloads the libzim binary, and executes it through some javascript bindings. Executing a binary is not possible in a (standard) browser environment, for security reasons.
To my knowledge, the only cross-platform and cross-browser way to execute a binary is to compile it with emscripten : either in webassembly or in asm.js.

The second step is to bind the recompiled C code with javascript.
nbind looks like a possible alternative to embind. It seems that it has been discussed with the emscripten team : https://github.com/emscripten-core/emscripten/issues/4770. It was rejected but it was 2 years ago...
In any case, nbind is not currently listed in their documentation : https://emscripten.org/docs/porting/connecting_cpp_and_javascript/index.html only mentions Embind and WebIDL
But, if it better suits our needs, why not

**Comment by mossroy** [comment]
I worked a bit on compiling kiwix-lib with emscripten (in branch https://github.com/mossroy/libzim_wasm/tree/kiwix-lib-compilation) : it's a long journey but I make progress (with many dirty hacks).
It forces to compile every dependency with emscripten too, so it takes time to make them work.

**Comment by mossroy** [comment]
I managed to compile the kiwix-lib with emscripten, and make a simple C call from javascript.
But I still can't make more than one C call : I probably do not do things right.

**Comment by mossroy** [comment]
When I make the second C++ call, it fails with `Assertion failed: you need to wait for the runtime to be ready (e.g. wait for main() to be called)` (even if main() has already been called).
I first thought it was because I was trying to keep the `Reader` instance in javascript, but it's not : I have the same issue if I re-create it each time (in C++).
This issue started when I used separate function calls (with embind). I tried to switch to WebIDL (in the more-WebIDL-binding-experiments branch) but did not manage to make it work because it tries to instantiate `Reader` through a constructor with no parameters (which does not exist in kiwix-lib). I doubt it is related to binding, in any case.

**Comment by mossroy** [comment]
Well, I have to admit I'm currently stuck.
I spent some time trying to make multiple C calls work, but did not find what's wrong.
I'm also not very efficient because I almost don't know how to code in C++ any more.
Some technical help would be appreciated on this issue.

**Comment by ISNIT0** [comment]
Would be happy to organise a call, it might be that what I've learned from `node-libzim` will be helpful?

**Comment by mossroy** [comment]
Thanks @ISNIT0 for the pointers you gave me during this call.

I'll try to sum up what we discussed :
- you faced technical issues when working on https://github.com/openzim/node-libzim (with nbind), and managed to solve them. So you suggest to try to switch to nbind instead of embind, in order to join forces on the same APIs. There are a few differences though : 
    - node-libzim uses the libzim, not kiwix-lib (but the principles should be the same)
    - node-libzim uses a binary version of the libzim. We need to use wasm instead (or asm.js). That's probably the main challenge
    - node-libzim uses typescript, but it should work in plain javascript too
- regarding the multiple c calls, it might come from the fact that I instantiate a class inside the C function. I should try to separate that into 2 functions : a first one that creates the instance (and keeps it in a global C variable), and a second one that uses it. I had tried that earlier today, but was unsuccessful because the C class does not have a default constructor with no parameters. In https://github.com/openzim/node-libzim/blob/master/src/ZimReader.cc, you worked around this issue by using a Wrapper class, that only keeps a reference to the instance we need : I should try the same approach

**Comment by mossroy** [comment]
I just tested to separate the code in 2 functions (initReader and getArticleCountFromReader) in https://github.com/mossroy/libzim_wasm/tree/multiple-c-calls. With the wrapper workaround, it compiles and runs, but there's the same error in the second call.

**Comment by ISNIT0** [comment]
Unfortunately I can't get the build to run on my Mac, so I can't play around with it :(

**Comment by mossroy** [comment]
@ISNIT0 : it's possible to build with Docker if it helps (see the README.md)

**Comment by mossroy** [comment]
This github issue has too much history : it's probably complicated to get into it.
I'll close it and create other smaller issues for what still needs to be done
